Add grpc example

thisthat opened the pull request.

thisthat committed:
Add build files for gRPC example

thisthat committed:
Add proto file, readme, and example code for gRPC

thisthat committed:
Improved naming and formatting

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,150 @@
+    TracerSdkFactory tracerFactory = OpenTelemetrySdk.getTracerFactory();
+    // Set to process the spans in memory
+    tracerFactory.addSpanProcessor(SimpleSpansProcessor.newBuilder(inMemexporter).build());
+    // Set the traces's name
```
```suggestion
    // Set the tracer's name
```

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,150 @@
+      logger.log(Level.WARNING, "RPC failed: {0}", e.getStatus());
+      return;
+    } finally {
+      span.end();
```
You should also set the span status to the result of the grpc call.

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,150 @@
+    span.setAttribute("MSG", "Will try to greet " + name + " ...");
+
+    // Set the context with the current span
+    tracer.withSpan(span);
```
You have to keep a reference to the returned `Scope` and close it once you're done.

thisthat committed:
parent dd8b598bff6447d80861fb9ff26136d622a602da
author Giovanni Liva <giovanni.liva@dynatrace.com> 1576073781 +0100
committer Giovanni Liva <giovanni.liva@dynatrace.com> 1576078475 +0100

Add Semantic Convention

arminru commented:
Please add span attributes according to the semantic conventions specified here:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-rpc.md

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,154 @@
+      } catch (StatusRuntimeException e) {
+        logger.log(Level.WARNING, "RPC failed: {0}", e.getStatus());
+        // TODO create mapping for io.grpc.Status<->io.opentelemetry.trace.Status
+        span.setStatus(Status.UNKNOWN);
```
```suggestion
        span.setStatus(Status.UNKNOWN.withDescription("GRPC status: " + e.getStatus()));
```

thisthat committed:
Add Semantic Convention

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,163 @@
+        tracer.spanBuilder("helloworld.Greeter/SayHello").setSpanKind(Span.Kind.CLIENT).startSpan();
+    // TODO provide attributes to Span.Builder
+    span.setAttribute("component", "grpc");
+    span.setAttribute("rpc.service", "helloworld.Greeter");
```
```suggestion
    span.setAttribute("rpc.service", "Greeter");
```
see https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-rpc.md#attributes

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldServer.java:
```diff
@@ -0,0 +1,184 @@
+              .startSpan();
+      // TODO provide attributes to Span.Builder
+      span.setAttribute("component", "grpc");
+      span.setAttribute("rpc.service", "helloworld.Greeter");
```
```suggestion
      span.setAttribute("rpc.service", "Greeter");
```

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,163 @@
+        return;
+      }
+    } finally {
+      span.end();
```
I'd personally move that to the inner try/catch block but it shouldn't matter.

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,163 @@
+    span.setAttribute("component", "grpc");
+    span.setAttribute("rpc.service", "helloworld.Greeter");
+    span.setAttribute("net.peer.name", this.hostname);
+    span.setAttribute("net.peer.port", this.port);
```
`this.hostname` and `this.port` is a bit confusing since `net.peer` describes the remote endpoint, i.e. the server to which this client is connecting. Something like `this.serverHostname` and `this.serverPort` would be clearer.

arminru commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldServer.java:
```diff
@@ -0,0 +1,184 @@
+      span.setAttribute("component", "grpc");
+      span.setAttribute("rpc.service", "helloworld.Greeter");
+      span.setAttribute("net.peer.name", "localhost");
+      span.setAttribute("net.peer.port", HelloWorldServer.this.port);
```
`net.peer` describes the remote endpoint, in this case the client.

You could set the server's port on `net.host.port`, however.

thisthat replied:
The following two specifications are misleading: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-rpc.md
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-span-general.md#net.name

arminru replied:
You're right. I opened https://github.com/open-telemetry/opentelemetry-specification/pull/386.

thisthat committed:
Update examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

thisthat committed:
Update examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldServer.java

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

thisthat committed:
Adapt example to support Semantic Conversion

arminru reviewed (approved):
Looks good now! :shipit:

z1c0 commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,163 @@
+  private final GreeterGrpc.GreeterBlockingStub blockingStub;
+
+  // OTel API
+  Tracer tracer;
```
Shouldn't those fields be `private` (`final`)?

z1c0 commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldClient.java:
```diff
@@ -0,0 +1,163 @@
+  // Export traces in memory
+  InMemorySpanExporter inMemexporter = InMemorySpanExporter.create();
+  // Share context via text headers
+  HttpTextFormat<SpanContext> textFormat;
```
In the "http sample" we removed this field in favor of `getHttpTextFormat` access when needed. Could we do this here as well?

z1c0 commented on examples/grpc/src/main/java/io/opentelemetry/example/HelloWorldServer.java:
```diff
@@ -0,0 +1,185 @@
+  private final int port = 50051;
+
+  // OTel API
+  Tracer tracer;
```
Same question as above in `HelloWorldClient`

z1c0 reviewed (approved):
LGTM! (couple of nits only)

thisthat committed:
Change exporter to LoggingExporter

thisthat committed:
Addressed the feedback of @bogdandrutu
